js/style/parse_color.js

```diff
@@ -0,0 +1,41 @@
+'use strict';
+
+var parseCSSColor = require('csscolorparser').parseCSSColor;
+var util = require('../util/util');
+
+var colorCache = {};
+
+function parseColor(input) {
+
+    if (colorCache[input]) {
+        return colorCache[input];
+
+    // RGBA array
+    } else if (Array.isArray(input)) {
+        return input;
+
+    // GL function
+    } else if (input && input.stops) {
+        return util.extend({}, input, {
+            stops: input.stops.map(function(step) {
+                return [step[0], parseColor(step[1])];
+            })
```

Comment:
This doesn't have to be a closure (created on every color) — function can be moved to top-level.